timings/test/timings_test.dart

```diff
@@ -0,0 +1,191 @@
+        });
+      });
+      await expectLater(future, throwsA(TypeMatcher<String>()));
+      expect(tracker.startTime, startTime);
+      expect(tracker.stopTime, time);
+      expect(tracker.duration, const Duration(seconds: 3));
+      additionalExpects();
+    });
+  }
+
+  group('SyncTimeTracker', () {
+    setUp(() {
+      tracker = SyncTimeTracker();
+    });
+
+    canHandleSync(() {});
+
+    test('Can not track async code', () async {
+      await scopedTrack(() => Future(() {
+            time = time.add(const Duration(seconds: 5));
+          }));
+      expect(tracker.startTime, startTime);
+      expect(tracker.stopTime, startTime);
+      expect(tracker.duration, const Duration(seconds: 0));
+    });
+  });
+
+  group('SyncTimeTracker.kindaAsync', () {
+    setUp(() {
+      tracker = SyncTimeTracker.kindaAsync();
+    });
+
+    canHandleSync(() {});
+
+    canHandleAsync(() {});
+
+    test('Can not distinguish own async code', () async {
+      var future = scopedTrack(() => Future(() {
+            time = time.add(const Duration(seconds: 5));
+          }));
+      time = time.add(const Duration(seconds: 10));
+      await future;
+      expect(tracker.startTime, startTime);
+      expect(tracker.stopTime, time);
+      expect(tracker.duration, const Duration(seconds: 15));
+    });
+  });
+
+  group('AsyncTimeTracker', () {
+    AsyncTimeTracker asyncTracker;
+    setUp(() {
+      tracker = asyncTracker = AsyncTimeTracker();
+    });
+
+    canHandleSync(() {
+      expect(asyncTracker.innerDuration, asyncTracker.duration);
+      expect(asyncTracker.slices.length, 1);
+    });
+
+    canHandleAsync(() {
+      expect(asyncTracker.innerDuration, asyncTracker.duration);
+      expect(asyncTracker.slices.length, greaterThan(1));
+    });
+
+    test('Can track complex async innerDuration', () async {
+      var future = scopedTrack(() {
+        time = time.add(const Duration(seconds: 1)); // Tracked sync
+        return Future(() {
+          time = time.add(const Duration(seconds: 2)); // Tracked async
+        }).then((_) async {
+          await Future.delayed(Duration.zero);
+          time = time.add(const Duration(seconds: 4)); // Tracked async, delayed
+        });
+      }).then((_) {
+        time = time.add(const Duration(seconds: 8)); // Async, after tracking
+      });
+      time = time.add(const Duration(seconds: 16)); // Sync, inside tracking
+      await Future.wait([
+        future,
+        // Async, inside tracking
+        Future.microtask(() => time = time.add(const Duration(seconds: 32)))
+      ]);
+      expect(asyncTracker.startTime, startTime);
+      expect(asyncTracker.stopTime.millisecondsSinceEpoch,
+          lessThan(time.millisecondsSinceEpoch));
+      expect(asyncTracker.duration, const Duration(seconds: 55));
+      expect(asyncTracker.innerDuration, const Duration(seconds: 7));
+      expect(asyncTracker.slices.length, greaterThan(1));
+    });
+  });
+}
```

Comment:
Related to my comment about nested performance zones earlier, we should add some tests for the nested zones to confirm their behavior as well.